### PR TITLE
Saul 220 - asset filter

### PR DESCRIPTION
**Pull request by Renkon**
missing update and tests

### CODE REVIEW

File: src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/_params.py [hunk 1]
```diff
@@ -114,6 +114,19 @@ def load_arguments(self, _):  # pylint: disable=too-many-locals, too-many-statem
...
+        c.argument('asset_name', help='The name of the asset.')
+        c.argument('filter_name', name_arg_type, id_part='child_name_1', help='The name of the asset filter.')
+        c.argument('start_timestamp', arg_group='Presentation Time Range', help='The absolute start time boundary.')
+        c.argument('end_timestamp', arg_group='Presentation Time Range', help='The absolute end time boundary.')
+        c.argument('presentation_window_duration', arg_group='Presentation Time Range', help='The relative to end sliding window.')
+        c.argument('live_backoff_duration', arg_group='Presentation Time Range', help='The relative to end right edge.')
+        c.argument('timescale', arg_group='Presentation Time Range', help='The time scale of time stamps.')
+        c.argument('force_end_timestamp', arg_group='Presentation Time Range', action='store_true', help='Indicates whether to force the existance of an end timestamp.')
+        c.argument('bitrate', help='The first quality bitrate.')
+        c.argument('tracks', help='The JSON representing the track selections. Use @{file} to load from a file.')
```
Comment (NahuelBerg) [inline src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/_params.py]: See argument in account-filter for reference

File: src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/_params.py [hunk 2]
```diff
@@ -114,6 +114,19 @@ def load_arguments(self, _):  # pylint: disable=too-many-locals, too-many-statem
                    help='The permissions to set on the SAS URL.')
         c.argument('expiry_time', expiry_arg_type, help="Specifies the UTC datetime (Y-m-d'T'H:M:S'Z') at which the SAS becomes invalid.")
 
+    with self.argument_context('ams asset-filter') as c:
+        c.argument('account_name', account_name_arg_type)
+        c.argument('asset_name', help='The name of the asset.')
```
Comment (damatojuan) [inline src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/_params.py]: this should have id_part='child_name_1'

File: src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/_params.py [hunk 3]
```diff
@@ -114,6 +114,19 @@ def load_arguments(self, _):  # pylint: disable=too-many-locals, too-many-statem
...
+        c.argument('filter_name', name_arg_type, id_part='child_name_1', help='The name of the asset filter.')
```
Comment (damatojuan) [inline src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/_params.py]: this should have id_part='child_name_2'

File: src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/_params.py (hunk 1 above)
Comment (damatojuan) [inline src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/_params.py]: Add allowed values

File: src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/commands.py [hunk 4]
```diff
@@ -88,6 +88,13 @@ def get_custom_sdk(custom_module, client_factory):
         g.custom_command('create', 'create_asset',
                          custom_command_type=get_custom_sdk('asset', get_assets_client))
 
+    with self.command_group('ams asset-filter', get_sdk('AssetFilters', get_asset_filters_client)) as g:
+        g.command('list', 'list')
+        g.command('show', 'get')
```
Comment (damatojuan) [inline src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/commands.py]: this should be a show_command

**Review by damatojuan — CHANGES_REQUESTED** [review]

File: src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/operations/asset_filter.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    if bitrate:
+        instance.first_quality = FirstQuality(bitrate=bitrate)
+
+    if any([start_timestamp, end_timestamp, presentation_window_duration,
+            live_backoff_duration, timescale, force_end_timestamp is not None]):
+
+        if instance.presentation_time_range is None:
+            if not all([start_timestamp, end_timestamp, presentation_window_duration,
+                        live_backoff_duration, timescale, force_end_timestamp is not None]):
+                raise CLIError('All parameters related to PresentationTimeRange must be set since object is null.')
```
Comment (NahuelBerg) [inline src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/operations/asset_filter.py]: The CLI user shouldn't understand what an 'object' is. Instead, I would write this:

> 'All parameters related to PresentationTimeRange must be set to create it.'

File: src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/tests/latest/test_ams_asset_filter_scenarios.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+            'amsname': amsname,
+            'storageAccount': storage_account_for_create,
+            'location': 'westus2',
+            'container': container
+        })
+
+        self.cmd('az ams account create -n {amsname} -g {rg} --storage-account {storageAccount} -l {location}')
+        self.cmd('az ams account storage add -a {amsname} -g {rg} -n {storageAccount}')
+
+        assetName = self.create_random_name(prefix='asset', length=12)
```
Comment (NahuelBerg) [inline src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/tests/latest/test_ams_asset_filter_scenarios.py]: You should use snake_case in python variables and camelCase for kwargs

**Review by NahuelBerg — CHANGES_REQUESTED** [review]

File: src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/tests/latest/test_ams_asset_filter_scenarios.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+            self.check('presentationTimeRange.liveBackoffDuration', '{liveBackoffDuration}'),
+            self.check('presentationTimeRange.presentationWindowDuration', '{presentationWindowDuration}'),
+            self.check('presentationTimeRange.startTimestamp', '{startTimestamp}'),
+            self.check('presentationTimeRange.timescale', '{timescale}'),
+            self.check('tracks[0].trackSelections[0].operation', 'Equal'),
+            self.check('tracks[0].trackSelections[0].property', 'FourCC'),
+            self.check('tracks[0].trackSelections[0].value', 'AVC1'),
+            self.check('tracks[1].trackSelections[0].operation', 'NotEqual'),
+            self.check('tracks[1].trackSelections[0].property', 'Unknown'),
+            self.check('tracks[1].trackSelections[0].value', 'AVC1'),
```
Comment (NahuelBerg) [inline src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/tests/latest/test_ams_asset_filter_scenarios.py]: If you look at the file, this assert shouldn't pass, it's EC-3, not AVC1

File: src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/operations/asset_filter.py (hunk 5 above)
Comment (damatojuan) [reply src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/operations/asset_filter.py]: what happens if we don't handle this? what is coming back from service?

File: src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/tests/latest/test_ams_asset_filter_scenarios.py (hunk 6 above)
Comment (damatojuan) [reply src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/tests/latest/test_ams_asset_filter_scenarios.py]: agree

File: src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/operations/asset_filter.py (hunk 5 above)
Comment (Renkon) [reply src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/operations/asset_filter.py]: it throws an unfriendly exception, although i could catch it and raise it as cli

File: src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/tests/latest/test_ams_asset_filter_scenarios.py (hunk 7 above)
Comment (Renkon) [reply src/command_modules/azure-cli-ams/azure/cli/command_modules/ams/tests/latest/test_ams_asset_filter_scenarios.py]: depending on whichever branch is merged first, either i or you will have to make changes

**Review by damatojuan — APPROVED** [review]

**Review by NahuelBerg — APPROVED** [review]

**Pull request merged**